swarmform/core/cluster.py

```diff
@@ -0,0 +1,435 @@
+                        p_parent.remove_child(parent.get_fw_id())
+                        # remove the parents from node
+                        parent.remove_parent(p_parent.get_fw_id())
+                        # Assign the parents of the nodes to the new clustered nodes
+                        if not is_parent_already_assigned(cluster, p_parent.get_fw_id()):
+                            cluster.add_parent(p_parent)
+                        # Assign new cluster as the child of parent nodes of the task
+                        if not is_child_already_assigned(p_parent, cluster.get_fw_id()):
+                            p_parent.add_child(cluster)
+                    for p_child in p_children:
+                        # remove the node as parent from node's child node
+                        p_child.remove_parent(parent.get_fw_id())
+                        # remove the children from node
+                        parent.remove_child(p_child.get_fw_id())
+                        # Assign the children of the nodes to the new clustered nodes
+                        if not is_child_already_assigned(cluster, p_child.get_fw_id()):
+                            cluster.add_child(p_child)
+                        # Assign new cluster as the parent of child nodes of the task
+                        if not is_parent_already_assigned(p_child, cluster.get_fw_id()):
+                            p_child.add_parent(cluster)
+                    # Delete the node from the WK
+                    wf.delete_node(parent.get_fw_id())
+                    # Assign a minus key to the refer the parallel running jobs
+                    key = -1*randrange(1, 100)
+                    # Set parallel nodes to the cluster
+                    cluster.set_parallel_ids([cluster.get_fw_ids_to_cluster_sequentially()[1],parent.get_fw_id()], key)
+                    # Set sequentially running nodes to the cluster
+                    cluster.set_sequential_ids([cluster.get_fw_ids_to_cluster_sequentially()[0], key])
+
+
+def wpa_clustering(workflow):
+
+    """
+    WPA clustering alogirthm
+
+    Args:
+        workflow (DAG)
+
+    Returns:
+        Clustered workflow DAG (DAG)
+    """
+    # Iterate the WF level by level
+    for level in range(workflow.get_height(), 1, -1):
+        cls_at_level = []
+        # Get the tasks at level
+        tasks_at_level = get_tasks_at_level(workflow, level)
+        # Sort the tasks by longest parent of the tasks
+        tasks_at_level_sorted = sort_tasks_by_longest_parent(tasks_at_level)
+        # Iterate the tasks of the level
+        for task in tasks_at_level_sorted:
+            # Assign parents of the task to the clusters
+            cls = assign_parent_to_clusters(task)
+            cls_info = {}
+            for cluster in cls:
+                cls_info = cluster.get_cluster_info()
+                temp = 0
+                for wfid in cls_info:
+                    if temp == 0:
```

Comment:
Do we need to have this if block? _temp_ is always 0 and checking whether it is 0 or not

Reply:
resolved that